src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb

```diff
@@ -372,21 +356,34 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 Case BoundKind.ByRefArgumentWithCopyBack
                     Return s_argumentMappings.GetValue(
                         argument,
-                        Function(argumentValue) New ByRefArgument(If(CUInt(index) < CUInt(parameters.Length), parameters(index), Nothing), DirectCast(argumentValue, BoundByRefArgumentWithCopyBack)))
+                        Function(argumentValue) New ByRefArgument(parameters(index), DirectCast(argumentValue, BoundByRefArgumentWithCopyBack)))
                 Case Else
-                    ' Apparently the VB bound trees don't encode named arguments, which seems unnecesarily lossy.
                     Return s_argumentMappings.GetValue(
                         argument,
                         Function(argumentValue)
-                            If index >= parameters.Length - 1 AndAlso parameters.Length > 0 AndAlso parameters(parameters.Length - 1).IsParamArray Then
-                                Return New Argument(ArgumentKind.ParamArray, parameters(parameters.Length - 1), argumentValue)
+                            Dim lastParameterIndex = parameters.Length - 1
+                            If index = lastParameterIndex AndAlso ParameterIsParamArray(parameters(lastParameterIndex)) Then
+                                ' TODO: figure out if this is true:
+                                '       a compiler generated argument for a ParamArray parameter is created iff 
+                                '       a list of arguments (including 0 argument) is provided for ParamArray parameter in source
+                                Dim kind = If(argument.WasCompilerGenerated AndAlso argument.Kind = BoundKind.ArrayCreation, ArgumentKind.ParamArray, ArgumentKind.Explicit)
+                                Return New Argument(kind, parameters(lastParameterIndex), argumentValue)
                             Else
-                                Return New Argument(ArgumentKind.Positional, If(CUInt(index) < CUInt(parameters.Length), parameters(index), Nothing), argumentValue)
+                                ' TODO: figure our if this is true:
+                                '       a compiler generated argument for an Optional parameter is created iff
+                                '       the argument is omitted from the source
+                                Dim kind = If(argument.WasCompilerGenerated, ArgumentKind.DefaultValue, ArgumentKind.Explicit)
+                                Dim parameter = parameters(index)
+                                Return New Argument(kind, parameter, argumentValue)
                             End If
                         End Function)
             End Select
         End Function
 
+        Private Shared Function ParameterIsParamArray(parameter As ParameterSymbol) As Boolean
+            Return If(parameter.IsParamArray AndAlso parameter.Type.Kind = SymbolKind.ArrayType, DirectCast(parameter.Type, ArrayTypeSymbol).IsSZArray, False)
+        End Function
```

Comment:
Why not simply `parameter.IsParamArray`? Please add a comment why more checks are needed.